lib/grpc/client/adapters/mint.ex

```diff
@@ -0,0 +1,227 @@
+defmodule GRPC.Client.Adapters.Mint do
+  @moduledoc """
+  A client adapter using mint
+  """
+
+  alias GRPC.{Channel, Credential}
+  alias GRPC.Client.Adapters.Mint.{ConnectionProcess, StreamResponseProcess}
+
+  @behaviour GRPC.Client.Adapter
+
+  @default_connect_opts [protocols: [:http2]]
+  @default_transport_opts [timeout: :infinity]
+
+  @impl true
+  def connect(%{host: host, port: port} = channel, opts \\ []) do
+    opts = Keyword.merge(@default_connect_opts, connect_opts(channel, opts))
+    Process.flag(:trap_exit, true)
+
+    channel
+    |> mint_scheme()
+    |> ConnectionProcess.start_link(host, port, opts)
+    |> case do
+      {:ok, pid} ->
+        {:ok, %{channel | adapter_payload: %{conn_pid: pid}}}
+
+      error ->
+        {:error, "An error happened while trying to opening the connection: #{inspect(error)}"}
+    end
+  catch
+    :exit, reason ->
+      {:error, "An error happened while trying to opening the connection: #{inspect(reason)}"}
+  end
+
+  @impl true
+  def disconnect(%{adapter_payload: %{conn_pid: pid}} = channel)
+      when is_pid(pid) do
+    :ok = ConnectionProcess.disconnect(pid)
+    {:ok, %{channel | adapter_payload: nil}}
+  end
+
+  def disconnect(%{adapter_payload: nil} = channel) do
+    {:ok, channel}
+  end
+
+  @impl true
+  def send_request(%{channel: %{adapter_payload: nil}}, _message, _opts),
+    do: raise("Can't perform a request without a connection process")
+
+  def send_request(stream, message, opts) do
+    {:ok, data, _} = GRPC.Message.to_data(message, opts)
+    do_request(stream, opts, data)
+  end
+
+  @impl true
+  def receive_data(stream, opts) do
+    cond do
+      success_bidi_stream?(stream) ->
+        do_receive_data(stream, :bidirectional_stream, opts)
+
+      success_server_stream?(stream) ->
+        do_receive_data(stream, :unary_request_stream_response, opts)
+
+      success_client_stream?(stream) ->
+        do_receive_data(stream, :stream_request_unary_response, opts)
+
+      success_unary_request?(stream) ->
+        do_receive_data(stream, :unary_request_response, opts)
```

Comment:
Ok, so I now get what the success_ functions mean.
I'd rather have a `success_response?` function and evaluate explicitly the grpc_type field here on the cond

Reply:
Ahh, right. This can definitely be simplified.